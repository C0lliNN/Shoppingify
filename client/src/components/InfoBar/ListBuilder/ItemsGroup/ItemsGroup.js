import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import * as variables from '../../../../helpers/style-constants';
import Button from '../../../UI/Button/Button';
import Checkbox from '../../../UI/Checkbox/Checkbox';

const Wrapper = styled.div`
  margin-top: 40px;
`;

const Category = styled.h5`
  font-family: ${variables.FONT_FAMILY};
  font-size: ${variables.FONT_SIZE_2};
  font-weight: 600;
  color: ${variables.COLORS.gray_3};
  margin-bottom: 20px;
`;

const ItemContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 15px auto;
`;

const ItemNameContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

const ItemName = styled.h5`
  font-family: ${variables.FONT_FAMILY};
  font-size: ${variables.FONT_SIZE_3};
  font-weight: 600;
  color: ${variables.COLORS.black_1};
`;

function ItemsGroup({ group, completing }) {
  return (
    <Wrapper>
      <Category>{group.category.name}</Category>
      {group.items.map((item) => (
        <ItemContainer key={item._id}>
          <ItemNameContainer>
            {!completing && <Checkbox />}
            <ItemName>{item.name}</ItemName>
          </ItemNameContainer>
          <Button btnType="outlined" variant="primary" disabled={!completing}>
            {item.quantity} pcs
          </Button>
        </ItemContainer>
      ))}
    </Wrapper>
  );
}

ItemsGroup.propTypes = {
  completing: PropTypes.bool.isRequired,
  group: PropTypes.shape({
    category: PropTypes.shape({
      _id: PropTypes.string,
      name: PropTypes.string,
    }),
    items: PropTypes.arrayOf(
      PropTypes.shape({
        _id: PropTypes.string,
        name: PropTypes.string,
        note: PropTypes.string,
        image: PropTypes.string,
        quantity: PropTypes.number,
        category: PropTypes.object,
      })
    ),
  }).isRequired,
};

export default ItemsGroup;
