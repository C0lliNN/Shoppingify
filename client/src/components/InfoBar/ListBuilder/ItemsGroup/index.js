import PropTypes from 'prop-types';
import React from 'react';
import Button from '../../../UI/Button/Button';
import Checkbox from '../../../UI/Checkbox';
import {
  Wrapper,
  Category,
  ItemContainer,
  ItemNameContainer,
  ItemName,
} from './styles';

function ItemsGroup({ group, completing }) {
  return (
    <Wrapper>
      <Category>{group.category.name}</Category>
      {group.items.map((item) => (
        <ItemContainer key={item._id}>
          <ItemNameContainer>
            {!completing && <Checkbox />}
            <ItemName>{item.name}</ItemName>
          </ItemNameContainer>
          <Button btnType="outlined" variant="primary" disabled={!completing}>
            {item.quantity} pcs
          </Button>
        </ItemContainer>
      ))}
    </Wrapper>
  );
}

ItemsGroup.propTypes = {
  completing: PropTypes.bool.isRequired,
  group: PropTypes.shape({
    category: PropTypes.shape({
      _id: PropTypes.string,
      name: PropTypes.string,
    }),
    items: PropTypes.arrayOf(
      PropTypes.shape({
        _id: PropTypes.string,
        name: PropTypes.string,
        note: PropTypes.string,
        image: PropTypes.string,
        quantity: PropTypes.number,
        category: PropTypes.object,
      })
    ),
  }).isRequired,
};

export default ItemsGroup;
