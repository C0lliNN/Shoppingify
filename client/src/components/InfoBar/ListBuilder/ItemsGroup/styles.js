import styled from 'styled-components';
import * as variables from '../../../../helpers/style-constants';

export const Wrapper = styled.div`
  margin-top: 40px;
`;

export const Category = styled.h5`
  font-family: ${variables.FONT_FAMILY};
  font-size: ${variables.FONT_SIZE_2};
  font-weight: 600;
  color: ${variables.COLORS.gray_3};
  margin-bottom: 20px;
`;

