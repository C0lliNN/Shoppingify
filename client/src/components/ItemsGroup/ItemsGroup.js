import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import * as variables from '../../helpers/style-constants';

const Title = styled.h3`
  font-family: ${variables.FONT_FAMILY};
  font-size: ${variables.FONT_SIZE_3};
  color: ${variables.COLORS.black_1};
  margin-top: 0px;
  margin-bottom: 15px;
  font-weight: 600;
`;

const Items = styled.div`
  display: flex;
  justify-content: space-around;
  align-items: center;
  flex-wrap: wrap;
`;

const Card = styled.div`
  background-color: ${variables.COLORS.white};
  border-radius: ${variables.BORDER_RADIUS_1};
  font-size: ${variables.FONT_SIZE_2};
  font-weight: 600;
  box-shadow: 0px 2px 12px rgba(0, 0, 0, 0.05);
  padding: 8px 20px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 80px;
  margin-bottom: 15px;
  @media (min-width: ${variables.MD_BREAK_POINT}px) {
    width: 120px;
    margin-bottom: 30px;
    font-size: ${variables.FONT_SIZE_3};
  }
`;

const Item = styled.span`
  cursor: pointer;
`;

const Icon = styled.i`
  font-size: ${variables.FONT_SIZE_3};
  color: #c1c1c4;
  cursor: pointer;
`;

function ItemsGroup({ category, items }) {
  return (
    <div style={{ marginTop: '50px' }}>
      <Title>{category}</Title>
      <Items>
        {items.map((item) => (
          <Card key={item.name}>
            <Item>{item.name}</Item>
            <Icon className="material-icons-round">add</Icon>
          </Card>
        ))}
      </Items>
    </div>
  );
}

ItemsGroup.propTypes = {
  category: PropTypes.string.isRequired,
  items: PropTypes.array.isRequired,
};

export default ItemsGroup;
