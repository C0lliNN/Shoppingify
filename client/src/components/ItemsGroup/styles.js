import styled from 'styled-components';
import * as variables from '../../helpers/style-constants';

export const Title = styled.h3`
  font-family: ${variables.FONT_FAMILY};
  font-size: ${variables.FONT_SIZE_3};
  color: ${variables.COLORS.black_1};
  margin-top: 0px;
  margin-bottom: 15px;
  font-weight: 600;
`;

export const Items = styled.div`
  display: flex;
  align-items: center;
  flex-wrap: wrap;
`;

export const Card = styled.div`
  background-color: ${variables.COLORS.white};
  border-radius: ${variables.BORDER_RADIUS_1};
  font-size: ${variables.FONT_SIZE_2};
  font-weight: 600;
  box-shadow: 0px 2px 12px rgba(0, 0, 0, 0.05);
  padding: 8px 20px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 80px;
  margin-bottom: 15px;
  margin-right: 5px;
  @media (min-width: ${variables.LG_BREAK_POINT}px) {
    width: 120px;
    margin-right: 15px;
    margin-bottom: 30px;
    font-size: ${variables.FONT_SIZE_3};
  }
`;

export const Item = styled.span`
  cursor: pointer;
`;

export const Icon = styled.i`
  font-size: ${variables.FONT_SIZE_3};
  color: #c1c1c4;
  cursor: pointer;
`;

export const Quantity = styled.span`
  font-size: ${variables.FONT_SIZE_1};
  color: #f9a10a;
  white-space: nowrap;
  margin-left: 12px;
`;
