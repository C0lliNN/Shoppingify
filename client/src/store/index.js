import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import authReducer from './reducers/auth';
import infoBarReducer from './reducers/info-bar';
import thunk from 'redux-thunk';

const rootReducer = combineReducers({
  auth: authReducer,
  infoBar: infoBarReducer,
});

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(thunk))
);

export default store;
