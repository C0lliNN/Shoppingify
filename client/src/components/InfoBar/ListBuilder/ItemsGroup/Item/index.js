import PropTypes from 'prop-types';
import React, { useState } from 'react';
import {
  ItemContainer,
  ItemNameContainer,
  ItemName,
  EditContainer,
} from './styles';
import Button from '../../../../UI/Button';
import Checkbox from '../../../../UI/Checkbox';
import { useDispatch } from 'react-redux';
import { decrementItem, incrementItem, removeItemFromList } from '../../../../../store/actions';

export default function Item({ item, completing }) {
  const [editing, setEditing] = useState(false);
  const dispatch = useDispatch();

  function handleIncrementItem() {
    dispatch(incrementItem(item));
  }

  function handleDecrementItem() {
    dispatch(decrementItem(item));
  }

  function handleRemoveItem() {
    dispatch(removeItemFromList(item));
  }

  function toggleEditing() {
    setEditing((prevState) => !prevState);
  }

  return (
    <ItemContainer key={item._id}>
      <ItemNameContainer>
        {completing && <Checkbox />}
        <ItemName>{item.name}</ItemName>
      </ItemNameContainer>
      {editing ? (
        <EditContainer>
          <button className="delete" onClick={handleRemoveItem}>
            <span className="material-icons-round">delete_outline</span>
          </button>
          <button className="icon" onClick={handleDecrementItem}>
            <span className="material-icons-round">remove</span>
          </button>
          <Button
            btnType="outlined"
            variant="primary"
            disabled={completing}
            onClick={toggleEditing}
          >
            {item.quantity} pcs
          </Button>
          <button className="icon" onClick={handleIncrementItem}>
            <span className="material-icons-round">add</span>
          </button>
        </EditContainer>
      ) : (
        <Button
          btnType="outlined"
          variant="primary"
          disabled={completing}
          onClick={toggleEditing}
        >
          {item.quantity} pcs
        </Button>
      )}
    </ItemContainer>
  );
}

Item.propTypes = {
  completing: PropTypes.any,
  item: PropTypes.shape({
    _id: PropTypes.string,
    name: PropTypes.string,
    quantity: PropTypes.number,
  }),
};
