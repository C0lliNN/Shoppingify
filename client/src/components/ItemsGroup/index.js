import PropTypes from 'prop-types';
import React from 'react';
import { showItemDetails, addItemToList } from '../../store/actions';
import { useDispatch } from 'react-redux';
import { Title, Items, Card, Item, Icon } from './styles';

function ItemsGroup({ category, items }) {
  const dispatch = useDispatch();

  return (
    <div style={{ marginTop: '50px' }}>
      <Title>{category.name}</Title>
      <Items>
        {items.map((item) => (
          <Card key={item._id}>
            <Item onClick={() => dispatch(showItemDetails(item))}>
              {item.name}
            </Item>
            <Icon
              className="material-icons-round"
              onClick={() => dispatch(addItemToList(item))}
            >
              add
            </Icon>
          </Card>
        ))}
      </Items>
    </div>
  );
}

ItemsGroup.propTypes = {
  category: PropTypes.shape({
    _id: PropTypes.string,
    name: PropTypes.string,
  }),
  items: PropTypes.array,
  showItemDetails: PropTypes.func,
};

export default ItemsGroup;
