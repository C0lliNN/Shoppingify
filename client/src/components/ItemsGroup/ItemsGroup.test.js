import React from 'react';
import { render } from '@testing-library/react';
import ItemsGroup from './ItemsGroup';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';

let errorMessage = null;
let category = null;
let items = null;

beforeEach(() => {
  console.error = jest.fn((message) => (errorMessage = message));

  category = 'Fruit';
  items = [{ name: 'Apple' }, { name: 'Banana' }, { name: 'Orange' }];
});

function exec() {
  return render(
    <Provider store={store}>
      <BrowserRouter>
        <ItemsGroup category={category} items={items} />
      </BrowserRouter>
    </Provider>
  );
}

describe('<ItemsGroup/>', () => {
  it('should console.error if a category is not defined', () => {
    category = undefined;

    exec();

    expect(console.error).toHaveBeenCalled();
    expect(errorMessage).toMatch(/category/i);
  });
  it('should console.error if items is not an array', () => {
    items = 12;

    expect(() => exec()).toThrowError();
  });
  it('should render correctly', () => {
    const { container } = exec();

    expect(container).toMatchSnapshot();
  });
  it('should display the category', () => {
    const { getByText } = exec();

    expect(getByText(category)).not.toBeNull();
  });
  it('should display the items', () => {
    const { getByText } = exec();

    items.forEach((item) => {
      expect(getByText(item.name)).not.toBeNull();
    });
  });
});
