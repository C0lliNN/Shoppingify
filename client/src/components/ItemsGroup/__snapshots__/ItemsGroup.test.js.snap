// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemsGroup/> should render correctly 1`] = `
<div>
  <div
    style="margin-top: 50px;"
  >
    <h3
      class="sc-AxjAm drHYgj"
    >
      Fruit
    </h3>
    <div
      class="sc-AxirZ gymQos"
    >
      <div
        class="sc-AxiKw eBVBFr"
      >
        <span>
          Apple
        </span>
        <i
          class="sc-AxhCb iPaXDK material-icons-round"
        >
          add
        </i>
      </div>
      <div
        class="sc-AxiKw eBVBFr"
      >
        <span>
          Banana
        </span>
        <i
          class="sc-AxhCb iPaXDK material-icons-round"
        >
          add
        </i>
      </div>
      <div
        class="sc-AxiKw eBVBFr"
      >
        <span>
          Orange
        </span>
        <i
          class="sc-AxhCb iPaXDK material-icons-round"
        >
          add
        </i>
      </div>
    </div>
  </div>
</div>
`;
